rfc/20240513-static-evaluation.md

```diff
@@ -0,0 +1,549 @@
+# Init-time static evaluation of constant variables and locals
+
+Issue: https://github.com/OpenTofu/OpenTofu/issues/1042
+
+As initially described in https://github.com/opentofu/opentofu/issues/1042, many users of OpenTofu expect to be able to use variables and locals in a variety of locations that are currently not supported.  Common examples include but are not limited to: module sources, provider assignments, backend configuration, encryption configuration.
+
+```hcl
+variable "accesskey" {
+    type = string
+}
+
+terraform {
+  backend "somebackend" {
+    accesskey = var.accesskey
+  }
+}
+```
+
+All of these examples are pieces of information that need to be known during `tofu init` and can not be integrated into the usual `tofu plan/apply` system. Init sets up a static understanding of the project's configuration that plan/apply can later operate on.
+
+## Proposed Solution
+
+In simple terms, the proposal is to enhance the configuration processing in OpenTofu to support evaluation of variable and local references that do not depend on any dynamic information (resources/data/providers/etc...).
+
+### User Documentation
+
+This proposal in and of itself does not explicitly add support for new user facing functionality in OpenTofu. It is designed to be a support system for the examples above.
+
+However, this "support system" will have to interact with the user directly. To properly define this support system, we will look at specific examples that will be built on top of this work.
+
+Note: All errors/warnings are simplified placeholders and would include better wording/formatting as well as source locations.
+
+#### Configurable Backend Examples
+
+Let's first look at a project's backend that depends on both variables and locals:
+
+```hcl
+variable "key" {
+    type = string
+}
+
+locals {
+    region = "us-east-1"
+    key_check = md5sum(var.key)
+}
+
+terraform {
+  backend "somebackend" {
+    region = local.region
+    key = var.key
+    key_check = local.key_check
+  }
+}
+```
+
+When first running `tofu init`, two errors will be produced:
```

Comment:
If I run a plan/apply and don't configure a variable value, tofu asks me to provide the variable's value in the command line.
Maybe we'll also want the same behavior for the `init` command?

Reply:
I think we should expand the explaination on line 60 below to cover this too.

Reply:
I don't know if doing so would be easy, as we'd probably need specific code for backend configuration references + module source references, etc.
But it is something that is very worth considering (maybe not for the first phase)

Reply:
Updated in [26374cd](https://github.com/opentofu/opentofu/pull/1649/commits/26374cd5c7309128052eb075036589dfcbc71671)